pallets/liquidity-pools/src/hooks.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright 2021 Centrifuge Foundation (centrifuge.io).
+// This file is part of Centrifuge chain project.
+
+// Centrifuge is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+use cfg_traits::{
+	investments::TrancheCurrency, liquidity_pools::OutboundQueue, StatusNotificationHook,
+};
+use cfg_types::{
+	domain_address::DomainAddress,
+	investments::{ExecutedForeignDecrease, ForeignInvestmentInfo},
+};
+use frame_support::{traits::fungibles::Mutate, transactional};
+use sp_runtime::{DispatchError, DispatchResult};
+use sp_std::marker::PhantomData;
+
+use crate::{pallet::Config, Message, MessageOf, Pallet};
+
+/// The hook struct which acts upon a finalized investment decrement.
+pub struct DecreaseInvestOrderHook<T>(PhantomData<T>);
+
+/// The hook struct which acts upon a finalized redemption collection.
+
+pub struct CollectRedeemHook<T>(PhantomData<T>);
+
+impl<T: Config> StatusNotificationHook for DecreaseInvestOrderHook<T>
+where
+	<T as frame_system::Config>::AccountId: Into<[u8; 32]>,
+{
+	type Error = DispatchError;
+	type Id = ForeignInvestmentInfo<T::AccountId, T::TrancheCurrency, ()>;
+	type Status = ExecutedForeignDecrease<T::Balance, T::CurrencyId>;
+
+	#[transactional]
+	fn notify_status_change(
+		id: ForeignInvestmentInfo<T::AccountId, T::TrancheCurrency, ()>,
+		status: ExecutedForeignDecrease<T::Balance, T::CurrencyId>,
+	) -> DispatchResult {
+		let ForeignInvestmentInfo {
+			id: investment_id,
+			owner: investor,
+			..
+		} = id;
+		let currency = Pallet::<T>::try_get_general_index(status.return_currency)?;
+		let wrapped_token = Pallet::<T>::try_get_wrapped_token(&status.return_currency)?;
+		let domain_address: DomainAddress = wrapped_token.into();
+
+		T::Tokens::burn_from(status.return_currency, &investor, status.amount_decreased)?;
+
+		let message: MessageOf<T> = Message::ExecutedDecreaseInvestOrder {
+			pool_id: investment_id.of_pool(),
+			tranche_id: investment_id.of_tranche(),
+			investor: investor.clone().into(),
+			currency,
+			currency_payout: status.amount_decreased,
+		};
+
+		T::OutboundQueue::submit(investor, domain_address.domain(), message)?;
```

Comment:
investor will need to pay if we use them as the sender. Not sure, if we should maybe use the treasury account here. In most case I would assume, that investors from other domain do not hold CFG in their derived address.

Reply:
Good point, we should and collect an appropriate amount from the investor beforehand as part of `DecreaseInvestOrder` and `CollectRedeem`.

Reply:
[ec7259d](https://github.com/centrifuge/centrifuge-chain/pull/1473/commits/ec7259d7dd82f1b578d94e5f6fb140b9490bf67d)